internal/client/api/leaves.go

```diff
@@ -0,0 +1,351 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package api
+
+import (
+	"slices"
+
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/database"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"github.com/tidwall/btree"
+)
+
+type leafSetItem[H comparable, N runtime.Number] struct {
+	hash   H
+	number N
+}
+
+// ImportOutcome privately contains the inserted and removed leaves after an import action.
+type ImportOutcome[H comparable, N runtime.Number] struct {
```

Comment:
So both the import action and the remove action both remove and insert items?

Reply:
Yup! Check `LeafSet.Import()`. It's pretty much exactly what you described.